shared/reducers/config.js

```diff
@@ -3,51 +3,107 @@ import logger from '../logger'
 import * as I from 'immutable'
 import * as Types from '../constants/types/config'
 import * as Constants from '../constants/config'
+import * as ChatConstants from '../constants/chat2'
 import * as ConfigGen from '../actions/config-gen'
 
 const initialState = Constants.makeState()
 
 export default function(
   state: Types.State = initialState,
-  action:
-    | ConfigGen.Actions
-    | ConfigGen.ChangedFocusPayload
-    | ConfigGen.ChangedActivePayload
-    | {type: 'remote:updateMenubarWindowID', payload: {id: number}}
+  action: ConfigGen.Actions | {type: 'remote:updateMenubarWindowID', payload: {id: number}}
 ): Types.State {
   switch (action.type) {
     case ConfigGen.resetStore:
-      return initialState
-        .set('readyForBootstrap', state.readyForBootstrap)
-        .set('menubarWindowID', state.menubarWindowID)
+      return initialState.merge({
+        appFocused: state.appFocused,
+        appFocusedCount: state.appFocusedCount,
+        configuredAccounts: state.configuredAccounts,
+        daemonHandshakeState: state.daemonHandshakeState,
+        daemonHandshakeWaiters: state.daemonHandshakeWaiters,
+        defaultUsername: state.defaultUsername,
+        logoutHandshakeWaiters: state.logoutHandshakeWaiters,
+        menubarWindowID: state.menubarWindowID,
+        pushLoaded: state.pushLoaded,
+        startupDetailsLoaded: state.startupDetailsLoaded,
+      })
+    case ConfigGen.restartHandshake:
+      return state.merge({
+        daemonError: null,
+        daemonHandshakeFailedReason: '',
+        daemonHandshakeRetriesLeft: Math.max(state.daemonHandshakeRetriesLeft - 1, 0),
+        daemonHandshakeState: 'starting',
+      })
+    case ConfigGen.startHandshake:
+      return state.merge({
+        daemonError: null,
+        daemonHandshakeFailedReason: '',
+        daemonHandshakeRetriesLeft: Constants.maxHandshakeTries,
+        daemonHandshakeState: 'starting',
+      })
+    case ConfigGen.logoutHandshake:
+      return state.merge({logoutHandshakeWaiters: I.Map()})
+    case ConfigGen.daemonHandshake:
+      return state.set('daemonHandshakeState', 'waitingForWaiters')
+    case ConfigGen.daemonHandshakeWait: {
+      if (state.daemonHandshakeState !== 'waitingForWaiters') {
+        throw new Error("Should only get a wait while we're waiting")
+      }
+
+      const oldCount = state.daemonHandshakeWaiters.get(action.payload.name, 0)
+      const newCount = oldCount + (action.payload.increment ? 1 : -1)
+      const newState =
+        newCount === 0
+          ? state.deleteIn(['daemonHandshakeWaiters', action.payload.name])
+          : state.setIn(['daemonHandshakeWaiters', action.payload.name], newCount)
+
+      if (action.payload.failedFatal) {
+        return newState.merge({
+          daemonHandshakeFailedReason: action.payload.failedReason || '',
+          daemonHandshakeRetriesLeft: 0,
+        })
+      } else {
+        // Keep the first error
```

Comment:
Why?

Reply:
i think its better to understand the first thing that went wrong to help debug whats going on, other errors are likely a symptom

Reply:
Sounds good